### ISSUE TITLE
[Question] Why `gamma * beta` stand for ` L2 in  LogisticRegression._NLL_grad

### ISSUE THREAD

**Issue by eromoe** [ddbourgin/numpy-ml#53]
Hello,  this is a great project , I am learning how to implement model without sklearn/tensorflow , it really help me a lot . 

I have a question on 
https://github.com/ddbourgin/numpy-ml/blob/4f37707c6c7c390645dec5a503c12a48e624b249/numpy_ml/linear_models/lm.py#L252

Since P-norm is defined as 
![image](https://user-images.githubusercontent.com/3938751/87290451-9957ed80-c530-11ea-98ae-0de91eed40ac.png)

`l1norms(self.beta)` means the sum of all absulote value of each element in self.beta .  I don't quite understand why the simple  `gamma * beta` stand for `L2  ?  

PS: May I ask what IDE and code document plugin you are using ? I see some annotation don't beyond to latex ,  it would be nice to see beautiful math symbols than raw latex :)

**Comment by ddbourgin** [comment]
For linear regression, the l2-regularization term is `gamma * np.sqrt(beta @ beta)`
The gradient of l2 penalty wrt beta is then simply `gamma * beta`

Keep in mind that `d_penality` is the gradient of the penalty term wrt the coefficients, not the penalty itself :)

I don't use a special IDE, unfortunately. the equations are formatted for display as Sphinx [reStructuredText](https://www.sphinx-doc.org/en/master/usage/restructuredtext/index.html). You can see the rendered equations in the online [documentation](https://numpy-ml.readthedocs.io/en/latest/), or build it yourself from the source in the `docs` directory. There may also be IDE plugins that will try to render them, but I am not aware of any :)

**Comment by eromoe** [comment]
Thank you for  reply .

From https://towardsdatascience.com/intuitions-on-l1-and-l2-regularisation-235f2db4c261


![image](https://user-images.githubusercontent.com/3938751/87842748-68393d80-c8e1-11ea-94fb-af4ebfde9faa.png)


 l1-regularization term is `gamma * np.absolute(beta)` 
 l2-regularization term is `gamma * np.power(np.sqrt(beta @ beta), 2)`  (I think you miswrote in previous comment )

The gradient of l1 penalty wrt beta is then simply `gamma  * np.sign(beta)`
The gradient of l2 penalty wrt beta is then simply `gamma * 2beta`  proportional to `gamma * beta` .

It looks strange  to me for the first sight  because you multiply `l1norm(beta)` in L1 case ,  since the gradient of l1 penalty is `gamma * np.sign(beta)` , this confused me .

**Comment by ddbourgin** [comment]
Whoops, yup, that's what I get for being hasty! The regularization penalty is [`(gamma / 2) * np.sqrt(beta @ beta) ** 2`](https://github.com/ddbourgin/numpy-ml/blob/4f37707c6c7c390645dec5a503c12a48e624b249/numpy_ml/linear_models/lm.py#L244), which gives a gradient of `gamma * beta`.

In the L1 case, I'd recommend explicitly writing down the L1 _penalty_ (not just the l1 norm) and then trying to derive the gradient wrt beta. It should quickly be clear why there is an `l1norm` term in the calc :)

**Comment by eromoe** [comment]
@ddbourgin Sorry but I don't quite understand why penalty in L1 case need square 
```
penalty = 0.5 * self.gamma * np.linalg.norm(self.beta, ord=1) ** 2  
```
All ariticles I saw was using  L1 term (penalty) 
![image](https://user-images.githubusercontent.com/3938751/88514973-8871a680-d01d-11ea-81c8-bd3128468f9d.png)
And the derivetive is +-\lambda  .
Now I am very confusing .

**Comment by ddbourgin** [comment]
Oh! I see what you're saying. You're right, the square of the L1 norm is not what we want. The proper L1 penalty is

`gamma * np.abs(beta).sum()`

which gives a gradient of

`gamma * np.sign(beta)`

I'll make a PR to fix this. Thank you very much for pointing this out :)
